File: src/luks2john.c
/* luks.c
 *
 * hashkill - a hash cracking tool
 * Copyright (C) 2010 Milen Rangelov <gat3way@gat3way.eu>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * gcc luks2john.c jumbo.c -o luks2john -lcrypto
 */


#if AC_BUILT
#include "autoconfig.h"
#else
#define _LARGEFILE64_SOURCE 1
#endif
#include "jumbo.h" // large file support
#include <stdio.h>
#if (!AC_BUILT || HAVE_UNISTD_H) && !_MSC_VER
#include <unistd.h>
#endif
#include <string.h>
#include "stdint.h"
#include <stdlib.h>
#include <sys/types.h>
#if !AC_BUILT || HAVE_FCNTL_H
#include <fcntl.h>
#endif
#if !AC_BUILT || HAVE_ARPA_INET_H
#include <arpa/inet.h>
#else
#include "johnswap.h"
#define ntohl JOHNSWAP
#endif
#include "params.h"
#include <openssl/bio.h>
#include <openssl/evp.h>

#define LUKS_MAGIC_L        6
#define LUKS_CIPHERNAME_L   32
#define LUKS_CIPHERMODE_L   32
#define LUKS_HASHSPEC_L     32
#define UUID_STRING_L       40
#define LUKS_DIGESTSIZE 20
#define LUKS_SALTSIZE 32
#define LUKS_NUMKEYS 8

static int inline_thr = MAX_INLINE_SIZE;
#define MAX_THR (LINE_BUFFER_SIZE / 2 - 2 * PLAINTEXT_BUFFER_SIZE)

/* taken from LUKS on disk format specification */
static struct luks_phdr {
	char magic[LUKS_MAGIC_L];
	uint16_t version;
	char cipherName[LUKS_CIPHERNAME_L];
	char cipherMode[LUKS_CIPHERMODE_L];
	char hashSpec[LUKS_HASHSPEC_L];
	uint32_t payloadOffset;
	uint32_t keyBytes;
	char mkDigest[LUKS_DIGESTSIZE];
	char mkDigestSalt[LUKS_SALTSIZE];
	uint32_t mkDigestIterations;
	char uuid[UUID_STRING_L];
	struct {
		uint32_t active;
		uint32_t passwordIterations;
		char passwordSalt[LUKS_SALTSIZE];
		uint32_t keyMaterialOffset;
		uint32_t stripes;
	} keyblock[LUKS_NUMKEYS];
} myphdr;

static unsigned char *cipherbuf;
static int afsize;
static unsigned int bestslot = 2000;


static int af_sectors(int blocksize, int blocknumbers)
{
	int af_size;

	af_size = blocksize * blocknumbers;
	af_size = (af_size + 511) / 512;
	af_size *= 512;
	return af_size;
}

static void print_hex(unsigned char *str, int len)
{
	int i;
	for (i = 0; i < len; ++i)
		printf("%02x", str[i]);
}

static int hash_plugin_parse_hash(char *filename)
{
	FILE *myfile;
	int cnt;
	int readbytes;
	unsigned int bestiter = 0xFFFFFFFF;

	myfile = jtr_fopen(filename, "rb");

	if (!myfile) {
		fprintf(stderr, "\n%s : %s!\n", filename, strerror(errno));
		return -1;
	}

	if (fread(&myphdr, sizeof(struct luks_phdr), 1, myfile) < 1) {
		fprintf(stderr, "%s : file opening problem!\n", filename);
		fclose(myfile);
		return -1;
	}

	if (strcmp(myphdr.magic, "LUKS\xba\xbe") != 0) {
		fprintf(stderr, "%s : not a LUKS file / disk\n", filename);
		fclose(myfile);
		return -2;
	}

	if (strcmp(myphdr.cipherName, "aes") != 0) {
		fprintf(stderr, "%s : Only AES cipher supported. Used cipher: %s\n",
			filename, myphdr.cipherName);
	}

	for (cnt = 0; cnt < LUKS_NUMKEYS; cnt++) {
		if ((ntohl(myphdr.keyblock[cnt].passwordIterations) < bestiter)
		    && (ntohl(myphdr.keyblock[cnt].passwordIterations) > 1) &&
		    (ntohl(myphdr.keyblock[cnt].active) == 0x00ac71f3)) {
			bestslot = cnt;
			bestiter =
			    ntohl(myphdr.keyblock[cnt].passwordIterations);
		}
	}
	if (bestslot == 2000)
		goto bad;

	afsize =
	    af_sectors(ntohl(myphdr.keyBytes),
	    ntohl(myphdr.keyblock[bestslot].stripes));

	fprintf(stderr, "Best keyslot [%d]: %d keyslot iterations, %d stripes, %d mkiterations\n", bestslot, ntohl(myphdr.keyblock[bestslot].passwordIterations),ntohl(myphdr.keyblock[bestslot].stripes),ntohl(myphdr.mkDigestIterations));
	fprintf(stderr, "Cipherbuf size: %d\n", afsize);

	/* common handling */
	cipherbuf = malloc(afsize);
	if (cipherbuf == NULL) {
		fprintf(stderr, "%s:%d: malloc failed\n", __FUNCTION__, __LINE__);
		exit(EXIT_FAILURE);
	}
	jtr_fseek64(myfile, ntohl(myphdr.keyblock[bestslot].keyMaterialOffset) * 512,
	SEEK_SET);
	readbytes = fread(cipherbuf, afsize, 1, myfile);
	if (readbytes < 0) {
		free(cipherbuf);
		fclose(myfile);
		goto bad;
	}

	if (afsize < inline_thr) {
		BIO *bio, *b64;
		fprintf(stderr, "Generating inlined hash!\n");
		printf("$luks$1$%zu$", sizeof(myphdr));
		print_hex((unsigned char *)&myphdr, sizeof(myphdr));
		printf("$%d$", afsize);
		/* base-64 encode cipherbuf */
		b64 = BIO_new(BIO_f_base64());
		bio = BIO_new_fp(stdout, BIO_NOCLOSE);
		bio = BIO_push(b64, bio);
		BIO_set_flags(bio, BIO_FLAGS_BASE64_NO_NL);
		BIO_write(bio, cipherbuf, afsize);
		if(BIO_flush(bio) <= 0) {
			fprintf(stderr, "%s : BIO_flush failed ;(\n", filename);
			fclose(myfile);
			return -3;
		}
		BIO_free_all(bio);
		printf("$");
		print_hex((unsigned char *)myphdr.mkDigest, LUKS_DIGESTSIZE);
		printf("\n");
		free(cipherbuf);
		goto good;
	}
	else {
		FILE *fp = jtr_fopen("dump", "wb");  // XXX make me unpredictable!
		fprintf(stderr, "Generating inlined hash with attached dump!\n");
		printf("$luks$0$%zu$", sizeof(myphdr));
		print_hex((unsigned char *)&myphdr, sizeof(myphdr));
		printf("$%d$", afsize);

		printf("%s$%s$", filename, "dump");
		print_hex((unsigned char *)myphdr.mkDigest, LUKS_DIGESTSIZE);
		printf("\n");

		fwrite(cipherbuf, afsize, 1, fp);
		free(cipherbuf);
		fclose(fp);

		goto good;
	}

good:
	fclose(myfile);
	return 0;
bad:
	printf("%s : parsing failed\n", filename);
	fclose(myfile);
	return 1;
}

static int usage(char *name)
{
	fprintf(stderr, "Usage: %s [-i <inline threshold>] [LUKS file(s) / disk(s)]\n"
	        "Default threshold is %d bytes (files smaller than that will be inlined)\n",
	        name, MAX_INLINE_SIZE);

	return EXIT_FAILURE;
}

int main(int argc, char **argv)
{
	int c;

	/* Parse command line */
	while ((c = getopt(argc, argv, "i:")) != -1) {
		switch (c) {
		case 'i':
			inline_thr = (int)strtol(optarg, NULL, 0);
			if (inline_thr > MAX_THR) {
				fprintf(stderr, "%s error: threshold %d, can't"
				        " be larger than %d\n", argv[0],
				        inline_thr, MAX_THR);
				return EXIT_FAILURE;
			}
			break;
		case '?':
		default:
			return usage(argv[0]);
		}
	}
	argc -= optind;
	if(argc == 0)
		return usage(argv[0]);
	argv += optind;

	while (argc--)
		hash_plugin_parse_hash(*argv++);

	return EXIT_SUCCESS;
}

File: run/luks2john.py
#!/usr/bin/env python

# This software is Copyright (c) 2016, Sanju Kholia <sanju.kholia at gmail.com>
# and it is hereby released to the general public under the following terms:
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted.
#
# luks2john.py is a rewrite of luks2john.c which was written by Milen Rangelov

import sys
import struct
from binascii import hexlify
import base64

PY3 = sys.version_info[0] == 3

"""
LUKS header, derived from LUKS on disk format specification

https://gitlab.com/cryptsetup/cryptsetup/wikis/Specification

static struct luks_phdr {
    char magic[LUKS_MAGIC_L];
    uint16_t version;
    char cipherName[LUKS_CIPHERNAME_L];
    char cipherMode[LUKS_CIPHERMODE_L];
    char hashSpec[LUKS_HASHSPEC_L];
    uint32_t payloadOffset;
    uint32_t keyBytes;
    char mkDigest[LUKS_DIGESTSIZE];
    char mkDigestSalt[LUKS_SALTSIZE];
    uint32_t mkDigestIterations;
    char uuid[UUID_STRING_L];
    struct {
        uint32_t active;
        uint32_t passwordIterations;
        char passwordSalt[LUKS_SALTSIZE];
        uint32_t keyMaterialOffset;
        uint32_t stripes;
    } keyblock[LUKS_NUMKEYS];
} myphdr;
"""

luks_header_fmt = '> 6s h 32s 32s 32s I I 20s 32s I 40s 384s'
luks_header_size = struct.calcsize(luks_header_fmt)
slot_fmt = "> I I 32s I I"
slot_size = 48


def af_sectors(blocksize, blocknumbers):
    af_size = blocksize * blocknumbers
    af_size = (af_size + 511) // 512
    af_size *= 512

    return af_size


def process_file(filename):
    bestiter = 0xFFFFFFFF
    bestslot = 2000
    LUKS_NUMKEYS = 8

    try:
        f = open(filename, "rb")
    except IOError:
        e = sys.exc_info()[1]
        sys.stderr.write("%s\n" % str(e))
        return

    myphdr = data = f.read(luks_header_size)
    if len(data) != luks_header_size:
        sys.stderr.write("%s : parsing failed\n" % filename)
        return -1

    data = struct.unpack(luks_header_fmt, data)
    (magic, version, cipherName, cipherMode, hashSpec, payloadOffset, keyBytes,
        mkDigest, mkDigestSalt, mkDigestIterations, uuid, slots) = data

    if magic != b"LUKS\xba\xbe":
        sys.stderr.write("%s : not a LUKS file / disk\n" % filename)
        return -2

    if version != 1:
        sys.stderr.write("%s : Only LUKS1 is supported. Used version: %d\n" %
                         (filename, version))
        return -2

    if not cipherName.startswith(b"aes\x00"):
        sys.stderr.write("%s : Only AES cipher supported. Used cipher: %s\n" %
                         (filename, cipherName))
        return -3

    if not cipherMode.startswith(b"cbc-essiv:sha256\x00"):
        sys.stderr.write("%s : Only cbc-essiv:sha256 mode is supported. Used mode: %s\n" %
                         (filename, cipherMode))
        return -4

    if not hashSpec.startswith(b"sha1\x00"):
        sys.stderr.write("%s : Only sha1 hash is supported. Used hash: %s\n" %
                         (filename, hashSpec))
        return -5

    # find the best slot
    for cnt in range(0, LUKS_NUMKEYS):
        data = slots[slot_size * cnt:slot_size * (cnt + 1)]
        data = struct.unpack(slot_fmt, data)
        (active, passwordIterations,
         passwordSalt, keyMaterialOffset, stripes) = data

        if passwordIterations < bestiter and passwordIterations > 1 \
                and active == 0x00ac71f3:
            bestslot = cnt
            bestiter = passwordIterations
            bestdata = data

    if bestslot == 2000:
        return -6

    afsize = af_sectors(keyBytes, stripes)
    (active, passwordIterations,
     passwordSalt, keyMaterialOffset, stripes) = bestdata

    sys.stderr.write("Best keyslot [%d]: %d keyslot iterations, %d stripes, %d mkiterations\n" %
                     (bestslot, passwordIterations, stripes,
                      mkDigestIterations))

    sys.stderr.write("Cipherbuf size: %d\n" % afsize)
    f.seek(keyMaterialOffset * 512, 0)
    cipherbuf = f.read(afsize)

    myphdr = hexlify(myphdr)
    cipherbuf = base64.b64encode(cipherbuf)
    mkDigest = hexlify(mkDigest)
    if PY3:
        myphdr = str(myphdr, 'ascii')
        cipherbuf = str(cipherbuf, 'ascii')
        mkDigest = str(mkDigest, 'ascii')
    sys.stdout.write("$luks$1$%d$%s" % (luks_header_size, myphdr))
    sys.stdout.write("$%d$%s$%s\n" % (afsize, cipherbuf, mkDigest))

    f.close()


if __name__ == "__main__":
    if len(sys.argv) < 2:
        sys.stderr.write("Usage: %s [LUKS file(s) / disk(s)]\n" % sys.argv[0])

    for i in range(1, len(sys.argv)):
        process_file(sys.argv[i])
